### PR TITLE
DumpModel_CAAS_fix

### PR DESCRIPTION
**Pull request by vinu2003**
## Description of change
Why is this change needed?
This PR address issue with dumpmodel command with CAAS models.
## QA steps
Execute dump-model command for CAAS models.
## Documentation changes
N/A
## Bug reference
N/A

### CODE REVIEW

File: state/application.go [hunk 1]
```diff
@@ -42,23 +43,24 @@ type Application struct {
...
+	CharmURL             *charm.URL   `bson:"charmurl"`
+	Channel              string       `bson:"cs-channel"`
+	CharmModifiedVersion int          `bson:"charmmodifiedversion"`
+	ForceCharm           bool         `bson:"forcecharm"`
+	Life                 Life         `bson:"life"`
+	UnitCount            int          `bson:"unitcount"`
+	RelationCount        int          `bson:"relationcount"`
+	Exposed              bool         `bson:"exposed"`
+	MinUnits             int          `bson:"minunits"`
+	Tools                *tools.Tools `bson:",omitempty"`
```
Comment (wallyworld) [inline state/application.go]: bson:"tools,omitempty"

File: state/application.go [hunk 2]
```diff
@@ -143,6 +145,17 @@ func (a *Application) Life() Life {
 	return a.doc.Life
 }
 
+// AgentTools returns the tools that the agent is currently running.
```
Comment (wallyworld) [inline state/application.go]: Here, use "operator" instead of "agent"

File: state/application.go [hunk 3]
```diff
@@ -143,6 +145,17 @@ func (a *Application) Life() Life {
...
+// It an error that satisfies errors.IsNotFound if the tools have not
+// yet been set.
+func (a *Application) AgentTools() (*tools.Tools, error) {
+	if a.doc.Tools == nil {
+		return nil, errors.NotFoundf("agent binaries for application %q", a)
```
Comment (wallyworld) [inline state/application.go]: operator image metadata for application

File: state/migration_export.go [hunk 4]
```diff
@@ -131,8 +131,10 @@ func (st *State) exportImpl(cfg ExportConfig) (description.Model, error) {
 	if err := export.modelUsers(); err != nil {
 		return nil, errors.Trace(err)
 	}
-	if err := export.machines(); err != nil {
-		return nil, errors.Trace(err)
+	if dbModel.Type() != ModelTypeCAAS {
```
Comment (wallyworld) [inline state/migration_export.go]: We really don't need this check. export.machines() is a no op for caas models. Having the check just adds complexity to the code.

File: state/migration_export.go [hunk 5]
```diff
@@ -673,6 +675,7 @@ type addApplicationContext struct {
 }
 
 func (e *exporter) addApplication(ctx addApplicationContext) error {
+	dbModel, err := e.st.Model()
```
Comment (wallyworld) [inline state/migration_export.go]: No need for this, the exporter has a model attribute to get the type

e.model.Type()

File: state/state.go [hunk 6]
```diff
@@ -1099,6 +1101,15 @@ func (st *State) AddApplication(args AddApplicationArgs) (_ *Application, err er
 		RelationCount: len(peers),
 		Life:          Alive,
 	}
+	if model.Type() == ModelTypeCAAS {
+		series, _ := series.HostSeries()
```
Comment (wallyworld) [inline state/state.go]: Series here is not the series of the host machine of the controller. It's the series of the caas cloud where for now is "kubernetes".
Same with arch

File: state/application.go [hunk 7]
```diff
@@ -143,6 +145,17 @@ func (a *Application) Life() Life {
 	return a.doc.Life
 }
 
+// AgentTools returns the tools that the operator is currently running.
+// It an error that satisfies errors.IsNotFound if the tools have not
+// yet been set.
+func (a *Application) AgentTools() (*tools.Tools, error) {
```
Comment (wallyworld) [inline state/application.go]: Also need the Set method

File: state/application.go [hunk 8]
```diff
@@ -143,6 +146,38 @@ func (a *Application) Life() Life {
...
+	if a.doc.Tools == nil {
+		return nil, errors.NotFoundf("operator image metadata for application %q", a)
+	}
+	tools := *a.doc.Tools
+	return &tools, nil
+}
+
+// SetAgentTools set the Tools value in applicationDoc
+func (a *Application) SetAgentTools(v version.Binary) (err error) {
```
Comment (wallyworld) [inline state/application.go]: SetAgentVersion
We need to implement the AgentTooler interface

File: state/application.go [hunk 9]
```diff
@@ -143,6 +146,38 @@ func (a *Application) Life() Life {
...
+	defer errors.DeferredAnnotatef(&err, "cannot set agent version for application %q", a)
+	if err = checkVersionValidity(v); err != nil {
+		return err
```
Comment (wallyworld) [inline state/application.go]: errors.Trace(err)

File: state/application.go [hunk 10]
```diff
@@ -143,6 +146,38 @@ func (a *Application) Life() Life {
...
+// yet been set.
+func (a *Application) AgentTools() (*tools.Tools, error) {
+	if a.doc.Tools == nil {
+		return nil, errors.NotFoundf("operator image metadata for application %q", a)
+	}
+	tools := *a.doc.Tools
+	return &tools, nil
+}
+
+// SetAgentVersion set the Tools value in applicationDoc
```
Comment (wallyworld) [inline state/application.go]: sets
Also, end comments with a "."

File: state/application_test.go [hunk 11]
```diff
@@ -3593,3 +3593,22 @@ func (s *ApplicationSuite) TestApplicationWaitAgentPresence(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(alive, jc.IsFalse)
 }
+
+/*func (s *ApplicationSuite) TestApplicationAgentTools(c *gc.C) {
```
Comment (wallyworld) [inline state/application_test.go]: Don't forget to add the tests here

File: state/migration_export.go [hunk 12]
```diff
@@ -822,6 +824,18 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
 		exUnit.SetConstraints(constraintsArgs)
 	}
 
+	// Set Tools for application - this is only for CAAS model
```
Comment (wallyworld) [inline state/migration_export.go]: "models."

File: state/migration_export_test.go [hunk 13]
```diff
@@ -420,6 +420,13 @@ func (s *MigrationExportSuite) assertMigrateApplications(c *gc.C, st *state.Stat
 		addr := network.NewScopedAddress("192.168.1.1", network.ScopeCloudLocal)
 		err = application.UpdateCloudService("provider-id", []network.Address{addr})
 		c.Assert(err, jc.ErrorIsNil)
+		agentTools := version.Binary{
```
Comment (wallyworld) [inline state/migration_export_test.go]: Adding tools is already done in the factory so no need to do it here

File: state/migration_export_test.go [hunk 14]
```diff
@@ -467,6 +474,10 @@ func (s *MigrationExportSuite) assertMigrateApplications(c *gc.C, st *state.Stat
 		c.Assert(addr.Scope(), gc.Equals, "local-cloud")
 		c.Assert(addr.Type(), gc.Equals, "ipv4")
 		c.Assert(addr.Origin(), gc.Equals, "provider")
+
+		tools, err := application.AgentTools()
+		c.Assert(err, jc.ErrorIsNil)
+		c.Assert(exported.Tools().Version(), gc.Equals, tools.Version)
 	} else {
 		c.Assert(exported.PodSpec(), gc.Equals, "")
 		c.Assert(exported.CloudService(), gc.IsNil)
```
Comment (wallyworld) [inline state/migration_export_test.go]: assert that AgentTools is nil

File: state/migration_export_test.go [hunk 15]
```diff
@@ -518,6 +529,16 @@ func (s *MigrationExportSuite) assertMigrateUnits(c *gc.C, st *state.State) {
 	s.primeStatusHistory(c, unit, status.Active, addedHistoryCount)
 	s.primeStatusHistory(c, unit.Agent(), status.Idle, addedHistoryCount)
 
+	if dbModel.Type() == state.ModelTypeIAAS {
```
Comment (wallyworld) [inline state/migration_export_test.go]: This code already exists in MakeUnit()

File: state/migration_export_test.go [hunk 16]
```diff
@@ -585,6 +606,12 @@ func (s *MigrationExportSuite) assertMigrateUnits(c *gc.C, st *state.State) {
 		c.Assert(addr.Type(), gc.Equals, "ipv4")
 		c.Assert(addr.Origin(), gc.Equals, "provider")
```
Comment (wallyworld) [inline state/migration_export_test.go]: Also assert that agent tools is nil

File: state/migration_import_test.go [hunk 17]
```diff
@@ -734,6 +734,7 @@ func (s *MigrationImportSuite) assertUnitsMigrated(c *gc.C, st *state.State, con
 		c.Assert(err, jc.ErrorIsNil)
 	}
 
+
```
Comment (wallyworld) [inline state/migration_import_test.go]: remove extra new line

File: testing/factory/factory.go [hunk 18]
```diff
@@ -439,6 +439,9 @@ func (factory *Factory) MakeApplication(c *gc.C, params *ApplicationParams) *sta
 // If params is not specified, defaults are used.
 // It returns the application and its password.
 func (factory *Factory) MakeApplicationReturningPassword(c *gc.C, params *ApplicationParams) (*state.Application, string) {
+	model, err := factory.st.Model()
```
Comment (wallyworld) [inline testing/factory/factory.go]: move this closer to where it's used

**Comment by vinu2003** [comment]
!!build!!

File: cmd/juju/model/dump_test.go [hunk 19]
```diff
@@ -75,6 +76,13 @@ func (s *DumpCommandSuite) TestDump(c *gc.C) {
 }
 
 func (s *DumpCommandSuite) TestDumpSimple(c *gc.C) {
+	err := s.store.UpdateModel("testing", "admin/mymodel", jujuclient.ModelDetails{
```
Comment (wallyworld) [inline cmd/juju/model/dump_test.go]: All these changes can be removed - dump_test.go should be unmodified

File: state/application_test.go [hunk 20]
```diff
@@ -1928,6 +1931,121 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+		CharmConfig: map[string]interface{}{
+			"foo": "bar",
+		},
+		ApplicationConfig: map[string]interface{}{
+			"app foo": "app bar",
+		},
+		ApplicationConfigFields: environschema.Fields{
+			"app foo": environschema.Attr{Type: environschema.Tstring}},
+	})
+	err := application.UpdateLeaderSettings(&goodToken{}, map[string]string{
```
Comment (wallyworld) [inline state/application_test.go]: Don't need at this extra stuff for this test - leader settings, metrics, annotations etc are purely optional parts of the application data model. We are not testing those aspects here.

File: state/application_test.go [hunk 21]
```diff
@@ -1928,6 +1931,121 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	err = application.SetMetricCredentials([]byte("sekrit"))
+	c.Assert(err, jc.ErrorIsNil)
+	dbModel, err := st.Model()
+	c.Assert(err, jc.ErrorIsNil)
+	err = dbModel.SetAnnotations(application, testAnnotations)
+	c.Assert(err, jc.ErrorIsNil)
+
+	if dbModel.Type() == state.ModelTypeCAAS {
```
Comment (wallyworld) [inline state/application_test.go]: We already know this is a CAAS model, but as before, we don't need these extra bits.

File: state/application_test.go [hunk 22]
```diff
@@ -1928,6 +1931,121 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+		caasModel, err := dbModel.CAASModel()
+		c.Assert(err, jc.ErrorIsNil)
+		err = caasModel.SetPodSpec(application.ApplicationTag(), "pod spec")
+		c.Assert(err, jc.ErrorIsNil)
+		addr := network.NewScopedAddress("192.168.1.1", network.ScopeCloudLocal)
+		err = application.UpdateCloudService("provider-id", []network.Address{addr})
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	model, err := st.Export()
```
Comment (wallyworld) [inline state/application_test.go]: No need to export anything - we are just testing the get/set tools behaviour of the application persistent state.

File: state/application_test.go [hunk 23]
```diff
@@ -1928,6 +1931,121 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
+
+	applications := model.Applications()
+	c.Assert(applications, gc.HasLen, 1)
+
+	agentTools := version.Binary{
```
Comment (wallyworld) [inline state/application_test.go]: These following lines are all we need.
We need to:
1. set the tools
2. refresh the application via application.Refresh()
(to ensure a reload from state)
3. test the tools

File: cmd/juju/model/dump_test.go [hunk 24]
```diff
@@ -54,6 +54,7 @@ func (s *DumpCommandSuite) SetUpTest(c *gc.C) {
 	s.store.Accounts["testing"] = jujuclient.AccountDetails{
 		User: "admin",
 	}
+
```
Comment (wallyworld) [inline cmd/juju/model/dump_test.go]: delete me

File: state/application_test.go [hunk 25]
```diff
@@ -1928,6 +1931,69 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
 }
 
+func (s *ApplicationSuite) TestAgentTools(c *gc.C) {
+	st := s.Factory.MakeModel(c, &factory.ModelParams{
+		Name: "caas-model",
+		Type: state.ModelTypeCAAS, CloudRegion: "<none>",
+		StorageProviderRegistry: factory.NilStorageProviderRegistry{}})
+	defer st.Close()
+	f := factory.NewFactory(st)
+	application := f.MakeApplication(c, &factory.ApplicationParams{
```
Comment (wallyworld) [inline state/application_test.go]: f.MakeApplication(c, nil)

File: state/application_test.go [hunk 26]
```diff
@@ -1928,6 +1931,69 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+}
+
+func (s *ApplicationSuite) TestSetAgentVersion(c *gc.C) {
+	st := s.Factory.MakeModel(c, &factory.ModelParams{
+		Name: "caas-model",
+		Type: state.ModelTypeCAAS, CloudRegion: "<none>",
+		StorageProviderRegistry: factory.NilStorageProviderRegistry{}})
+	defer st.Close()
+	f := factory.NewFactory(st)
+	application := f.MakeApplication(c, &factory.ApplicationParams{
```
Comment (wallyworld) [inline state/application_test.go]: f.MakeApplication(c, nil)

File: state/application_test.go [hunk 27]
```diff
@@ -1928,6 +1931,69 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+	})
+	agentTools := version.Binary{
+		Number: jujuversion.Current,
+		Arch:   arch.HostArch(),
+		Series: application.Series(),
+	}
+
+	tools, err := application.AgentTools()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(tools.Version, gc.Equals, agentTools)
```
Comment (wallyworld) [inline state/application_test.go]: jc.DeepEquals

File: state/application_test.go [hunk 28]
```diff
@@ -1928,6 +1931,69 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+
+	err := application.SetAgentVersion(agentTools)
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = application.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+
+	tools, err := application.AgentTools()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(tools.Version, gc.Equals, agentTools)
```
Comment (wallyworld) [inline state/application_test.go]: jc.DeepEquals

File: state/application_test.go [hunk 29]
```diff
@@ -1928,6 +1931,69 @@ func (s *ApplicationSuite) TestAddCAASUnit(c *gc.C) {
...
+			"foo": "bar",
+		},
+		ApplicationConfig: map[string]interface{}{
+			"app foo": "app bar",
+		},
+		ApplicationConfigFields: environschema.Fields{
+			"app foo": environschema.Attr{Type: environschema.Tstring}},
+	})
+	agentTools := version.Binary{
+		Number: jujuversion.Current,
```
Comment (wallyworld) [inline state/application_test.go]: version.MustParse("2.99.0")

(or some other fake version number)
We want it to be different to version.Current

**Comment by vinu2003** [comment]
!!build!!

**Comment by vinu2003** [comment]
$$merge$$

**Pull request merged**